File: aoc/2020/aoc202015.hs
import Data.Map(Map,empty,findWithDefault,fromList,insert)

play :: [Int] -> [Int]
play = playStart 1 empty
  where
    playStart turn ages [] = error "playStart"
    playStart turn ages [n] = playNext turn ages n
    playStart turn ages (n:ns) = n : playStart (turn+1) (insert n turn ages) ns
    playNext turn ages n = n : playNext (turn+1) (insert n turn ages) (turn - findWithDefault turn n ages)

veSeq :: [Int] -> Int -> Int
veSeq start i
  | i <= length start = start !! (i-1)
  | otherwise = run (fromList (zip (init start) [1..])) (length start) (last start)
  where
    run history j n
      | j == i = n
      | otherwise = run (insert n j history) (j+1) next
      where next = j - findWithDefault j n history

test :: ()
test
  | play [0,3,6] !! 2019 /= 436 = error "a"
  | play [1,3,2] !! 2019 /= 1 = error "b"
  | play [2,1,3] !! 2019 /= 10 = error "c"
  | play [1,2,3] !! 2019 /= 27 = error "d"
  | play [2,3,1] !! 2019 /= 78 = error "e"
  | play [3,2,1] !! 2019 /= 438 = error "f"
  | play [3,1,2] !! 2019 /= 1836 = error "g"
  | otherwise = ()

part1 :: [Int] -> Int
part1 input = play input !! 2019

-- Get stack overflow for part 2.

-- https://oeis.org/A181391
-- No indication of any faster algorithms.

File: aoc/2020/aoc202015.go
package main

import (
	"os"
	"strconv"
)

const nmax = 30000000

var history [nmax]int

func clearHistory() {
	for i := range history {
		history[i] = 0
	}
}

func play(n int, start []int) int {
	i := 0
	num := 0
	for _, next := range start {
		history[num] = i
		num = next
		i++
	}
	for i < n {
		next := history[num]
		history[num] = i
		if next != 0 {
			num = i - next
		} else {
			num = 0
		}
		i++
	}
	return num
}

func main() {
	if len(os.Args) < 2 {
		println(play(2020, []int{0,3,6}), 436)
		clearHistory()
		println(play(2020, []int{1,3,2}), 1)
		clearHistory()
		println(play(2020, []int{2,1,3}), 10)
		clearHistory()
		println(play(2020, []int{1,2,3}), 27)
		clearHistory()
		println(play(2020, []int{2,3,1}), 78)
		clearHistory()
		println(play(2020, []int{3,2,1}), 438)
		clearHistory()
		println(play(2020, []int{3,1,2}), 1836)
		clearHistory()
		println(play(30000000, []int{0,3,6}), 175594)
		clearHistory()
		println(play(30000000, []int{1,3,2}), 2578)
		clearHistory()
		println(play(30000000, []int{2,1,3}), 3544142)
		clearHistory()
		println(play(30000000, []int{1,2,3}), 261214)
		clearHistory()
		println(play(30000000, []int{2,3,1}), 6895259)
		clearHistory()
		println(play(30000000, []int{3,2,1}), 18)
		clearHistory()
		println(play(30000000, []int{3,1,2}), 362)
	} else {
		start := []int{}
		for _, arg := range os.Args[1:] {
			n, err := strconv.Atoi(arg)
			if err != nil {
				panic(err.Error())
			}
			start = append(start, n)
		}
		println(play(2020, start))
		clearHistory()
		println(play(30000000, start))
	}
}
